### DISCUSSION TITLE
Javinizer via Docker on Synology DSM

### DISCUSSION THREAD

**Discussion by MattePainting** [javinizer/Javinizer#162] [Q&A] [answered]
Thanks to everyone working on this!

Does anyone have the Javinizer web GUI running via Docker on a Synology NAS? I'm trying to get a container up and running, but I'm struggling. My knowledge level is low in general, and nonexistent when it comes to Docker.

I'm using the Synology DSM web interface, and I have Docker installed with the the latest Javinizer image downloaded from Docker Hub. My trouble is with what settings I need to adjust before launching the container, and then how to access it once it's launched.

I will need to change the port from 5000 to something else, as 5000 is what's already in use for Synology DSM itself. Am I correct in assuming the container port should be kept as 5000, and the local port adjusted to something not already in use? Or do I have this backwards?

I'm also not entirely sure what I should be using for file/folder mount paths. I created a folder for Javinizer at the following path on my NAS: `/volume1/docker/javinizer` Would the mount path for that folder be `/data`? The GitBook guide suggests mounting the settings file as well. Does this file path: `/volume1/docker/javinizer/jvSettings.json` with this mount path: `/home/Javinizer/src/Javinizer/jvSettings.json` make sense?

I plan on using a pretty simple path for the input and output once I have the container running: `/volume1/JAV`

I've attached screenshots of what the Advanced Settings pages look like in Docker when launching a Javinizer container in Synology DSM, if it helps visualize my questions. 

![Synology_DiskStation_Docker_Javinizer_Container_Settings_0](https://user-images.githubusercontent.com/45110649/102677949-2b64c180-4173-11eb-9d83-af215f3c5693.png)
![Synology_DiskStation_Docker_Javinizer_Container_Settings_1](https://user-images.githubusercontent.com/45110649/102677954-3586c000-4173-11eb-9c7f-dd95b3668893.png)
![Synology_DiskStation_Docker_Javinizer_Container_Settings_2](https://user-images.githubusercontent.com/45110649/102677958-37508380-4173-11eb-82f3-1b36eda79436.png)
![Synology_DiskStation_Docker_Javinizer_Container_Settings_3](https://user-images.githubusercontent.com/45110649/102677959-39b2dd80-4173-11eb-9952-dc19d99df520.png)
![Synology_DiskStation_Docker_Javinizer_Container_Settings_4](https://user-images.githubusercontent.com/45110649/102677961-3ae40a80-4173-11eb-8115-c2b8f8afd8a2.png)
![Synology_DiskStation_Docker_Javinizer_Container_Settings_5](https://user-images.githubusercontent.com/45110649/102677963-3cadce00-4173-11eb-89f9-9f129b90512a.png)
![Synology_DiskStation_Docker_Javinizer_Container_Settings_6](https://user-images.githubusercontent.com/45110649/102677965-3e779180-4173-11eb-82eb-34a402efe94c.png)

I realize these questions are more about Docker more than Javinizer, but I hope the discussion might help others using Synology.

**Comment by jvlflame** [comment, marked as answer]
> I will need to change the port from 5000 to something else, as 5000 is what's already in use for Synology DSM itself. Am I correct in assuming the container port should be kept as 5000, and the local port adjusted to something not already in use? Or do I have this backwards?

This is correct. Your container port will need to remain as 5000 as that is what the PowerShell Universal instance inside the container is exposing. Your local (host) port is what needs to be changed since that is conflicting with an existing exposed Synology port.

> I'm also not entirely sure what I should be using for file/folder mount paths. I created a folder for Javinizer at the following path on my NAS: /volume1/docker/javinizer Would the mount path for that folder be /data? The GitBook guide suggests mounting the settings file as well. Does this file path: /volume1/docker/javinizer/jvSettings.json with this mount path: /home/Javinizer/src/Javinizer/jvSettings.json make sense?

Yep, that's correct.
`/volume1/docker/javinizer` -> `/data`
`/volume1/docker/javinizer/jvSettings.json` -> `/home/Javinizer/src/Javinizer/jvSettings.json`

> I plan on using a pretty simple path for the input and output once I have the container running: /volume1/JAV

In that case you can mount it to any desired folder inside the container.

For example:
`/volume1/JAV` -> `/mnt`

In this case, to access your `/volume1/JAV` folder, you would set your `location.input` setting within Javinizer to use `/mnt`. My suggesting to create two folders within the `/volume1/JAV` folder called `input` and `output`, and then set your `location.output` setting to `/mnt/output`.

**Reply by MattePainting** [reply]
Thanks for the reply. I have the container up and running now, but I'm not able to access it at the local port I set in my container settings. I feel like I'm overlooking something obvious, but I will continue troubleshooting.

**Reply by jvlflame** [reply]
I'm not very familiar with Synology's Docker UI, but I don't see where you have your host port defined in the screenshots you provided. 

Edit: Actually, I see the local port being set as `Auto`. 
Try both ways:
- Set container port as 20000 and local port as 5000.
- Set container port as 5000 and local port as 20000.

Try accessing the container at `http://<ip address>:20000` with both configurations.

**Reply by MattePainting** [reply]
Setting the local port as 5000 doesn't work due to the conflict with Synology DSM. I tried setting the local port to 20000 instead of using the auto assigned one, but I still can't access the container at `http://<ip address>:20000`. I tried a few other options for local port with no success. Here's the 'Overview' page for the container with the Port Settings displayed.
![Synology_DiskStation_Docker_Javinizer_Container_Overview](https://user-images.githubusercontent.com/45110649/102700742-4b04f400-421e-11eb-929f-bf89b3479031.png)

I'm getting the impression that even though the container is running, what's in the container is not. This is all I see on the 'Process' tab inside the container details:
![Synology_DiskStation_Docker_Javinizer_Container_Process](https://user-images.githubusercontent.com/45110649/102700645-4724a200-421d-11eb-9cfb-941a1a3026b9.png)

And the 'Log' tab:
![Synology_DiskStation_Docker_Javinizer_Container_Log](https://user-images.githubusercontent.com/45110649/102700648-4e4bb000-421d-11eb-9603-605099eb9423.png)

It seems like there's a step between launching the container and getting PowerShell Universal running inside of it that I'm completely overlooking.

**Reply by jvlflame** [reply]
I actually recall another user had trouble setting up Javinizer using the Synology Docker GUI, and had to eventually run the docker command via the terminal. I believe it may have been an issue with the "Execution Command" not running correctly. I would suggest you delete the existing container and SSH into your Synology or access the terminal and run `docker run --name javinizer -p 20000:5000 -d jvlflame/javinizer:latest`.

If it still doesn't work, get the logs of the running container with `docker logs javinizer` and we can troubleshoot from there.

**Reply by MattePainting** [reply]
Deleting the container and then starting over using the docker command via SSH did the trick, I'm seeing the PowerShell Universal login page at `http://<ip address>:20000` now.

Thanks for the help!

**Reply by ghost** [reply]
> I actually recall another user had trouble setting up Javinizer using the Synology Docker GUI, and had to eventually run the docker command via the terminal. I believe it may have been an issue with the "Execution Command" not running correctly. I would suggest you delete the existing container and SSH into your Synology or access the terminal and run `docker run --name javinizer -p 20000:5000 -d jvlflame/javinizer:latest`.
> 
> If it still doesn't work, get the logs of the running container with `docker logs javinizer` and we can troubleshoot from there.

thanks for the suggestion, i've run it via ssh and now successfully installed javinizer. 
for the moment i already try to sort demo folder (around 1-2 title), and after click sort it move and create new folder to `output` folder with new .nfo files.

the thing is, i already made a bunch of directories with Actress Name (root folder), and Title (subfolder). I just want the Javinizer made the .nfo files in existing folder, or creating .nfo files without moving the movie file itself. i tried to change the configuration on Javinizer settings but no luck (still moved the movie files)

![image](https://user-images.githubusercontent.com/4795465/103452943-77813b80-4d07-11eb-8f64-e19264c76f17.png)

from what i read on guidebook https://docs.jvlflame.net/ , you can change configuration via cli. can you please tell me how to do it on Synology NAS? is it via container page? because i can't type anything here
![image](https://user-images.githubusercontent.com/4795465/103452969-d8a90f00-4d07-11eb-89bc-966e8c6550c2.png)

or run it via javinizer console?
![image](https://user-images.githubusercontent.com/4795465/103452987-1312ac00-4d08-11eb-881d-45d1642cd716.png)

please kindly help, i'm at loss here

**Reply by jvlflame** [reply]
@lord2bo `sort.movetofolder` off should do what you're looking for. Not sure why it isn't.. but one more thing you might also try is using the "Update" checkbox on the sort tab. That will create metadata files without moving files.
![image](https://user-images.githubusercontent.com/17932235/103453074-dbb1ea00-4c8a-11eb-804c-3c4903f3f29a.png)

Also, you don't need to do Javinizer -OpenSettings within the console. That's built into the settings page. Just click on the "JSON" button to edit it directly.

**Reply by ghost** [reply]
> @lord2bo `sort.movetofolder` off should do what you're looking for. Not sure why it isn't.. but one more thing you might also try is using the "Update" checkbox on the sort tab. That will create metadata files without moving files.
> ![image](https://user-images.githubusercontent.com/17932235/103453074-dbb1ea00-4c8a-11eb-804c-3c4903f3f29a.png)
> 
> Also, you don't need to do Javinizer -OpenSettings within the console. That's built into the settings page. Just click on the "JSON" button to edit it directly.

thankyou for the prompt response, it works perfectly.
it is nice to see all my collections organized. lol. thanks again

**Comment by mitrestreet** [comment]
The above has been really helpful, as someone with a Synology NAS wishing to run this application. However, I have hit an impasse at the next step, and would appreciate any help that can be offered.

I have the Docker container running, am able to login to the Powershell Universal configuration pages, but after following the instructions in the GitBook, am unable to start the dashboard. The error I get on the powershell dashboard is "Failed to start dashboard: Request failed with status code 404"

Looking at the container log, this section seems relevant:
fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
      An unhandled exception has occurred while executing the request.
System.Exception: Failed to start dashboard.
   at Universal.Server.Services.DashboardProxy.Start(Dashboard dashboard) in D:\a\universal\universal\src\Universal.Server\Services\DashboardProxy.cs:line 118
   at Universal.Server.Services.DashboardManager.Start(Dashboard dashboard) in D:\a\universal\universal\src\Universal.Server\Services\DashboardManager.cs:line 65
   at UniversalAutomation.DashboardController.Start(Int64 id) in D:\a\universal\universal\src\Universal.Server\Controllers\DashboardController.cs:line 163
   at lambda_method(Closure , Object , Object[] )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Universal.Server.Middleware.RoutingMiddleware.Invoke(HttpContext httpContext, IPolicyEvaluator policyEvaluator) in D:\a\universal\universal\src\Universal.Server\Middleware\RoutingMiddleware.cs:line 88
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at AspNetCoreRateLimit.RateLimitMiddleware`1.Invoke(HttpContext context) in D:\a\universal\universal\src\AspNetCoreRateLimit\Middleware\RateLimitMiddleware.cs:line 109
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)

What have I done wrong?

**Reply by MattePainting** [reply]
Hmm, I would double-check that you followed the steps for importing the dashboard correctly in Powershell Universal. I just spun up a new container to see if I got any errors, and I didn't encounter any problems importing or accessing the Javinizer dashboard afterwards: https://docs.jvlflame.net/installation/install-javinizer-web-gui#import-the-javinizer-dashboard

**Reply by mitrestreet** [reply]
Thank you for responding so quickly. After I posted that, the dashboard worked when I clicked start, but came up with "hello world!". I pasted in the code from main.ps1 to the code window accessible from the dashboard entry's info button, and now have something akin to what is expected - I now need to work out how to make settings persistent and pick up the correct directories. Can I use volumes instead of bind mounts? I ask as I figure I can use the Docker GUI in DSM to add these (as I can't work out the syntax to bind mount at the command line.

**Reply by MattePainting** [reply]
I set the bind mount / volume settings via the Docker application in Synology DSM after I got the container running using command line via SSH. I didn't bother looking up the syntax for doing that part on command line. You should see the Javinizer container in the Docker application, even if you created it via command line. You can stop the container and then click 'Edit' to change the volume settings, and then start it again when you're finished.

Here's what my current volume settings look like in the Docker application, if it's helpful:
![Volume_Settings](https://user-images.githubusercontent.com/45110649/103300736-98772300-49cd-11eb-91c7-46373d415623.png)

**Reply by jvlflame** [reply]
> Thank you for responding so quickly. After I posted that, the dashboard worked when I clicked start, but came up with "hello world!". I pasted in the code from main.ps1 to the code window accessible from the dashboard entry's info button, and now have something akin to what is expected - I now need to work out how to make settings persistent and pick up the correct directories. Can I use volumes instead of bind mounts? I ask as I figure I can use the Docker GUI in DSM to add these (as I can't work out the syntax to bind mount at the command line.

EDIT: My bad, didn't fully read that you figured it out on your own! Disregard this post 👍 . I'll leave this here for any others who may find this information valuable.

Just a heads up, if you delete the Javinizer dashboard in the PowerShell Universal GUI after importing it, it deletes the actual Javinizer dashboard files. I'm assuming that's why it's showing up with the default dashboard page.

 You'll either have to recreate the container from scratch or you can open the "Info" page of the dashboard you created, click the 'Edit' button, and copy in the code from [here](https://github.com/jvlflame/Javinizer/blob/master/dashboard/main.ps1). This is what it should look like: 
![image](https://user-images.githubusercontent.com/17932235/103318589-c6676200-49e3-11eb-84c0-fbcd7b19454b.png)

**Reply by mitrestreet** [reply]
Well, I got much further, but can't seem to keep the thing running. I added the volumes as per your settings, with one difference (based on updated documentation): the application settings point /home/data in the container.

I have been able to successfully use the system to tag files. However, I am not sure what triggers the issue, but after around 30 minutes of uptime, or possibly a successful run, I start getting errors on the Settings page of the dashboard:

Cannot process argument transformation on parameter 'Checked'. Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, such as $True, $False, 1 or 0.

This error shows up in one or more places, eventually replacing all the boxes in the Scrapers and Sort sections of the Settings page. I have noticed that any parameters in the settings file have been overwritten and replaced with "null". For example:

"location.input": null

Not sure why this is happening, or how to fix it..,

**Reply by jvlflame** [reply]
That's strange... the settings file should not be affected in any way while idle or after performing a successful sort.

It's only affected by the checkboxes on the sort page as well as the apply and json buttons on the settings page. I would suggest to use the json editor on the settings page and paste in the default configuration. https://raw.githubusercontent.com/jvlflame/Javinizer/master/src/Javinizer/jvSettings.json

If it continues happening or you find a way to reproduce it, please let me know.

**Reply by ypyoo** [reply]
I have same problem...Is there any progress or solution found yet?

**Reply by jvlflame** [reply]
> I have same problem...Is there any progress or solution found yet?

I don't have a SynoIogy, but I haven't been able to reproduce this issue on my Javinizer docker container. Is it just happening randomly for you or do you have a way to consistently reproduce the error?

**Reply by ypyoo** [reply]
Thank you for kind reply...
It happen consistently in my Synology NAS

I attached my setting 

Again I appreciate for your support and great project!
![스크린샷 2021-01-13 오전 11 49 05](https://user-images.githubusercontent.com/47057552/104400544-f7c06000-5595-11eb-8e60-5e47f2b5353c.png)
![스크린샷 2021-01-13 오전 11 48 51](https://user-images.githubusercontent.com/47057552/104400549-fa22ba00-5595-11eb-8b8b-3013b9cdd1bb.png)
![스크린샷 2021-01-13 오전 11 48 32](https://user-images.githubusercontent.com/47057552/104400551-fb53e700-5595-11eb-8b51-9ecc02eb42c2.png)

**Reply by ghost** [reply]
> Thank you for kind reply...
> It happen consistently in my Synology NAS
> 
> I attached my setting
> 
> Again I appreciate for your support and great project!
> ![스크린샷 2021-01-13 오전 11 49 05](https://user-images.githubusercontent.com/47057552/104400544-f7c06000-5595-11eb-8e60-5e47f2b5353c.png)
> ![스크린샷 2021-01-13 오전 11 48 51](https://user-images.githubusercontent.com/47057552/104400549-fa22ba00-5595-11eb-8b8b-3013b9cdd1bb.png)
> ![스크린샷 2021-01-13 오전 11 48 32](https://user-images.githubusercontent.com/47057552/104400551-fb53e700-5595-11eb-8b51-9ecc02eb42c2.png)

try look into your json settings, make sure it is filled with true or false expression.
go to settings -> json (click json button to edit)
mine came out the same errors like you, turns out there is null expression there.

![Capture](https://user-images.githubusercontent.com/4795465/104812505-3c1d5b80-5835-11eb-9f34-3730a9743278.JPG)

hope it helps

**Reply by ypyoo** [reply]
Thank you! I solved problem and enjoying Javnizer! 
The problem was setting.jason file I downloaded it and replace.

I think Mac OS is not good for this solution. Windows' chrome browser is showing better performance. Safari shows error sometimes...

Thank you~have great time!

**Comment by ukiya21** [comment]
Hello. Hoped can help.. Currently I'm using DS918+. Wanted to know how to update to newer version without removing the old version. I tried to type -UpdateModule on the terminal but nothing happened. TQ

**Comment by vuongtt92** [comment]
Hi all,
I try to install javinizer via Docker on Synology DSM. I run ssh into synology and run this command `XXX@DS918:~$ docker run --name javinizer -p 8600:8600 -d javinizer/javinizer:latest
`
But I've got this error message:
```
docker: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create?name=javinizer": dial unix /var/run/docker.sock: connect: permission denied.
See 'docker run --help'.
```

Could you please help? I'm not familiar with code and Docker (very entry level) Thank you everyone
